File: 11-Data_structures_graphs/graph.rb
# frozen_string_literal: true

class Graph
  def initialize
    @number_of_nodes = 0
    @adjacent_list = {}
  end

  def add_vertex(node)
    @adjacent_list[node] = []
    @number_of_nodes += 1
  end

  def add_edge(node1, node2)
    @adjacent_list[node1] << node2
    @adjacent_list[node2] << node1
  end

  def show_connections
    all_nodes = @adjacent_list.keys

    all_nodes.each do |node|
      node_connections = @adjacent_list[node]
      connections = ''

      node_connections.each do |vertex|
        connections += vertex
      end

      puts "#{node} --> #{connections}"
    end
  end
end

graph = Graph.new
graph.add_vertex('0')
graph.add_vertex('1')
graph.add_vertex('2')
graph.add_vertex('3')
graph.add_vertex('4')
graph.add_vertex('5')
graph.add_vertex('6')
graph.add_edge('3', '1')
graph.add_edge('3', '4')
graph.add_edge('4', '2')
graph.add_edge('4', '5')
graph.add_edge('1', '2')
graph.add_edge('1', '0')
graph.add_edge('0', '2')
graph.add_edge('6', '5')
graph.show_connections

File: 11-Data_structures_graphs/graph.js
class Graph {
  constructor() {
    this.numberOfNodes = 0;
    this.adjancentList = {};
  }

  addVertex(node) {
    this.adjancentList[node] = [];
    this.numberOfNodes++;
  }

  addEdge(node1, node2) {
    // undirected graph
    this.adjancentList[node1].push(node2);
    this.adjancentList[node2].push(node1);
  }

  showConnections() {
    const allNodes = Object.keys(this.adjancentList);

    for (let node of allNodes) {
      let nodeConnections = this.adjancentList[node];
      let connections = '';
      let vertex;

      for (vertex of nodeConnections) {
        connections += vertex + ' ';
      }

      console.log(node + ' --> ' + connections);
    }
  }
}

const graph = new Graph();
graph.addVertex('0');
graph.addVertex('1');
graph.addVertex('2');
graph.addVertex('3');
graph.addVertex('4');
graph.addVertex('5');
graph.addVertex('6');
graph.addEdge('3', '1');
graph.addEdge('3', '4');
graph.addEdge('4', '2');
graph.addEdge('4', '5');
graph.addEdge('1', '2');
graph.addEdge('1', '0');
graph.addEdge('0', '2');
graph.addEdge('6', '5');
graph.showConnections();
